### PR TITLE
Terraform to perform CRUD operations for LB resource

### PR DESCRIPTION
**Pull request by Pushpalatha555**
**VALIDATED THE LB FOR CREATE/DELETE/UPDATE**

CREATE
UI-CREATE
UPDATE
UI-UPDATE
DELETE
UI-DELETE

### CODE REVIEW

File: examples/data-sources/hpegl_vmaas_tier1_router/data-source.tf [hunk 1]
```diff
@@ -0,0 +1,5 @@
+# (C) Copyright 2021 Hewlett Packard Enterprise Development LP
+
+data "hpegl_vmaas_tier1_router" "tier1_router" {
```
Comment (manjunath-batakurki) [inline examples/data-sources/hpegl_vmaas_tier1_router/data-source.tf:3]: is there a data source exclusively for hpegl_vmaas_tier1_router ? I don't think so. you can add under **https://github.com/HewlettPackard/hpegl-vmaas-terraform-resources/tree/master/examples/data-sources/hpegl_vmaas_router** and change the name to nsx-tier0 and nsx-tier1 for the file names.

File: examples/resources/hpegl_vmaas_load_balancer/nsx_t_lb.tf [hunk 2]
```diff
@@ -1,15 +1,17 @@
...
 
-resource "hpegl_vmaas_load_balancer" "tf_load_balancer" {
-  name  = "lab-1"       
-  description  = "creating load balancer for test"
-  enabled      =    true        
-  visibility   = "public"
-  resource_permission {
+resource "hpegl_vmaas_load_balancer" "tf_lb" {
+  name  = "lab-2"       
+  description  = "CREATE load balancer for test"
```
Comment (manjunath-batakurki) [inline examples/resources/hpegl_vmaas_load_balancer/nsx_t_lb.tf:5]: please use correct names here(name, description), later these examples would be used to write documentation for this resource.

File: examples/resources/hpegl_vmaas_load_balancer/nsx_t_lb.tf [hunk 3]
```diff
@@ -1,15 +1,17 @@
...
+  enabled      =  true 
+  group_access {
     all = true
   }
   config {
     admin_state = true
-    tier1  = "/infra/tier-1s/26cdb82e-0057-4461-ad4d-cddd61d77b1f"
+    size = "SMALL"
+    log_level = "INFO"
+    tier1_gateways  = data.hpegl_vmaas_tier1_router.tier1_router.provider_id
```
Comment (manjunath-batakurki) [inline examples/resources/hpegl_vmaas_load_balancer/nsx_t_lb.tf:14]: hpegl_vmaas_tier1_router doesn't exist

File: internal/cmp/client.go [hunk 4]
```diff
@@ -86,6 +87,7 @@ func NewClient(client *apiClient.APIClient, cfg apiClient.Configuration) *Client
 			&apiClient.ProvisioningAPIService{Client: client, Cfg: cfg}),
 		CloudFolder:   newCloudFolder(&apiClient.CloudsAPIService{Client: client, Cfg: cfg}),
 		DSRouter:      newRouterDS(&apiClient.RouterAPIService{Client: client, Cfg: cfg}),
+		DSTier1Router: newTier1RouterDS(&apiClient.RouterAPIService{Client: client, Cfg: cfg}),
```
Comment (manjunath-batakurki) [inline internal/cmp/client.go:90]: not needed

File: internal/cmp/lb_profiles.go [hunk 5]
```diff
@@ -58,7 +58,7 @@ func (lb *loadBalancerProfile) Create(ctx context.Context, d *utils.Data, meta i
 				RequestHeaderSize:      d.GetInt("request_header_size"),
 				ResponseHeaderSize:     d.GetInt("response_header_size"),
 				ResponseTimeout:        d.GetInt("response_timeout"),
-				HTTPIdleTimeoutName:    d.GetInt("http_idle_timeout"),
```
Comment (manjunath-batakurki) [inline internal/cmp/lb_profiles.go:61]: will review this file when Profiles are implemented.

File: internal/cmp/loadbalancer.go [hunk 6]
```diff
@@ -37,60 +37,51 @@ func (lb *loadBalancer) Read(ctx context.Context, d *utils.Data, meta interface{
...
+		nameKey: nsxt,
+	})
 
-	for i, n := range types.LoadBalancerTypes {
+	for i, n := range allTypes.LoadBalancerTypes {
 		if n.Name == nsxt {
-			createReq.NetworkLoadBalancer.Type = types.LoadBalancerTypes[i].Name
-			createReq.NetworkLoadBalancer.NetworkServerID = 1
+			createReq.NetworkLoadBalancer.Type = allTypes.LoadBalancerTypes[i].Name
+			createReq.NetworkLoadBalancer.NetworkServerID = networkServerID
```
Comment (manjunath-batakurki) [inline internal/cmp/loadbalancer.go:74]: Can you explain this ? I hope u are getting the LB Type here. refer tier0 gateway type or nsx network type.

File: internal/cmp/tier1_router.go [hunk 7]
```diff
@@ -0,0 +1,46 @@
+// (C) Copyright 2021 Hewlett Packard Enterprise Development LP
+
```
Comment (manjunath-batakurki) [inline internal/cmp/tier1_router.go:2]: this fle is not needed

File: internal/resources/constants.go [hunk 8]
```diff
@@ -21,6 +21,7 @@ const (
 	DSNetworkInterface = "hpegl_vmaas_network_interface"
 	DSCloudFolder      = "hpegl_vmaas_cloud_folder"
 	DSRouter           = "hpegl_vmaas_router"
+	DSTier1Router      = "hpegl_vmaas_tier1_router"
```
Comment (manjunath-batakurki) [inline internal/resources/constants.go:24]: not needed

File: internal/resources/constants.go [hunk 9]
```diff
@@ -34,6 +35,7 @@ const (
 	ResInstanceClone              = "hpegl_vmaas_instance_clone"
 	ResNetwork                    = "hpegl_vmaas_network"
 	ResRouter                     = "hpegl_vmaas_router"
+	ResTier1Router                = "hpegl_vmaas_tier1_router"
```
Comment (manjunath-batakurki) [inline internal/resources/constants.go:38]: not needed

File: internal/resources/resource_loadbalancer.go [hunk 10]
```diff
@@ -19,15 +19,15 @@ func LoadBalancer() *schema.Resource {
 				Required:    true,
 				Description: "Network loadbalancer name",
 			},
-			"type": {
+			"nsx_t": {
```
Comment (manjunath-batakurki) [inline internal/resources/resource_loadbalancer.go:22]: can this be named as lb_type ?

File: internal/resources/resource_loadbalancer.go [hunk 11]
```diff
@@ -37,32 +37,47 @@ func LoadBalancer() *schema.Resource {
...
+							Optional:    true,
 							Default:     true,
+							Description: "Pass `true` to allow access to all groups.",
+						},
+						"sites": {
+							Type:        schema.TypeList,
 							Optional:    true,
-							Description: "If `true` then resource_permission rule will be active/enabled.",
+							Description: "List of sites/groups",
+							Elem: &schema.Resource{
```
Comment (manjunath-batakurki) [inline internal/resources/resource_loadbalancer.go:59]: please validate setting all to false and assigning to only one group.

File: pkg/resources/registration.go [hunk 12]
```diff
@@ -36,11 +36,11 @@ func (r Registration) SupportedDataSources() map[string]*schema.Resource {
 		resources.DSNetworkInterface: resources.NetworkInterfaceData(),
 		resources.DSCloudFolder:      resources.CloudFolderData(),
 		resources.DSRouter:           resources.RouterData(),
+		resources.DSTier1Router:      resources.Tier1RouterData(),
```
Comment (manjunath-batakurki) [inline pkg/resources/registration.go:39]: not needed.

File: internal/cmp/client.go (hunk 4 above)
Comment (Pushpalatha555) [reply internal/cmp/client.go:90]: done

File: internal/resources/data_source_tier1_router.go [hunk 13]
```diff
@@ -0,0 +1,78 @@
+// (C) Copyright 2021 Hewlett Packard Enterprise Development LP
```
[the remark this answers is not in the document]
Comment (Pushpalatha555) [reply internal/resources/data_source_tier1_router.go:1]: done

File: examples/resources/hpegl_vmaas_load_balancer/nsx_t_lb.tf (hunk 3 above)
Comment (Pushpalatha555) [reply examples/resources/hpegl_vmaas_load_balancer/nsx_t_lb.tf:14]: updated

File: internal/cmp/tier1_router.go (hunk 7 above)
Comment (Pushpalatha555) [reply internal/cmp/tier1_router.go:2]: deleted

File: internal/resources/constants.go (hunk 8 above)
Comment (Pushpalatha555) [reply internal/resources/constants.go:24]: removed

File: internal/resources/constants.go (hunk 9 above)
Comment (Pushpalatha555) [reply internal/resources/constants.go:38]: removed

File: internal/resources/resource_loadbalancer.go (hunk 10 above)
Comment (Pushpalatha555) [reply internal/resources/resource_loadbalancer.go:22]: done

File: pkg/resources/registration.go (hunk 12 above)
Comment (Pushpalatha555) [reply pkg/resources/registration.go:39]: done

File: examples/resources/hpegl_vmaas_load_balancer/nsx_t_lb.tf (hunk 2 above)
Comment (Pushpalatha555) [reply examples/resources/hpegl_vmaas_load_balancer/nsx_t_lb.tf:5]: done

File: internal/resources/resource_loadbalancer.go (hunk 11 above)
Comment (Pushpalatha555) [reply internal/resources/resource_loadbalancer.go:59]: done

File: internal/resources/resource_loadbalancer.go [hunk 14]
```diff
@@ -74,18 +89,19 @@ func LoadBalancer() *schema.Resource {
...
 							Description: `In Filter. Supported Values are "DEBUG", "INFO", "WARNING", "ERROR", "CRITICAL", "ALERT", "EMERGENCY"`,
 						},
-						"tier1": {
+						"tier1_gateways": {
 							Type:        schema.TypeString,
-							Optional:    true,
-							Default:     true,
+							Required:    true,
 							Description: "Network Loadbalancer NSX-T tier1 gateway",
 						},
```
[the remark this answers is not in the document]
Comment (Pushpalatha555) [reply internal/resources/resource_loadbalancer.go:106]: done

File: internal/cmp/constants.go [hunk 15]
```diff
@@ -24,5 +24,8 @@ const (
 	tier1GatewayType             = "NSX-T Tier-1 Gateway"
 	routerFirewallExternalPolicy = "GatewayPolicy"
 
+	// load balancer consts
+	networkServerID = 1
```
[the remark this answers is not in the document]
Comment (Pushpalatha555) [reply internal/cmp/constants.go:28]: removed

File: internal/cmp/loadbalancer.go [hunk 16]
```diff
@@ -37,60 +39,80 @@ func (lb *loadBalancer) Read(ctx context.Context, d *utils.Data, meta interface{
...
+
+func (lb *loadBalancer) Create(ctx context.Context, d *utils.Data, meta interface{}) error {
+	setMeta(meta, lb.lbClient.Client)
+	var createReq models.CreateLoadBalancerRequest
+	if err := tftags.Get(d, &createReq.NetworkLoadBalancer); err != nil {
+		return err
+	}
+
+	// align createReq and fill json related fields
+	if err := lb.loadBalancerAlignRequest(ctx, meta, &createReq); err != nil {
```
Comment (manjunath-batakurki) [inline internal/cmp/loadbalancer.go:106]: is this function not needed in update API ?

File: internal/resources/data_source_loadbalancer.go [hunk 17]
```diff
@@ -1,88 +0,0 @@
-// (C) Copyright 2022 Hewlett Packard Enterprise Development LP
-
-package resources
-
-import (
-	"context"
-
```
Comment (manjunath-batakurki) [inline internal/resources/data_source_loadbalancer.go:7]: I'm assuming you will add DS for LB in another review

**Review by manjunath-batakurki — COMMENTED** [review]
Looks Good, few minor comments.

File: internal/resources/data_source_loadbalancer.go (hunk 17 above)
Comment (Pushpalatha555) [reply internal/resources/data_source_loadbalancer.go:7]: yes

File: internal/cmp/loadbalancer.go (hunk 16 above)
Comment (Pushpalatha555) [reply internal/cmp/loadbalancer.go:106]: not needed

**Review by manjunath-batakurki — APPROVED** [review]
LGTM

**Review by gandharvas — APPROVED** [review]

**Pull request merged**